### ISSUE TITLE
Unintuitive interpretation of some interval operator phrases

### ISSUE THREAD

**Issue by brynrhodes** [cqframework/clinical_quality_language#109]
There is an ambiguity in the specification related to the interpretation of timing relationships. Specifically, given this example:

    define Interval1: Interval[@2012-01-01T09:00:00.0, @2012-01-01T10:30:00.0]
    define Interval2: Interval[@2012-01-01T11:00:00.0, @2012-01-01T18:00:00.0]
    define Within1: Interval1 ends 1 hour or less before start of Interval2

The specification implies in the definition of timing relationships (2.5.5.3) that this should return `true`, but then describes in 5.5.3.2 that it should be interpreted as a duration between that would return `false`.

The problem is that there are two boundaries involved in the comparison "X 1 hour or less before Y":

1) The boundary specified by the "before", which is exclusive by the definition of the "before" operator.
2) The boundary specified by the relative qualifier, "1 hour or less", which is clearly inclusive.

In this particular case, those boundaries meet and so the interpretation is ambiguous. (In the case of "1 hour or more" the boundaries don't meet, so there is no ambiguity and you get the intuitive behavior). 

The translator should be updated to provide the intuitively correct behavior by having the inclusive boundary of the relative qualifier override the exclusive boundary of the comparison operator.

An STU comment has been submitted to the specification to clarify the ambiguity:
http://www.hl7.org/dstucomments/showdetail_comment.cfm?commentid=1336

**Comment by cmoesel** [comment]
If the inclusive boundary overrides the exclusive boundary, then does that mean that if `X` was _at the same exact time_ as `Y`, then `X 1 hour or less before Y` would be `true`?  I think that's _better_ than the current behavior, but now it violates the `before` part of the relationship (since happening at the same time is _not_ before).

A more correct translation might be a logical `and` of `X before Y` and `hours between X and Y <= 1`.  Its more complex, but I _think_ more semantically correct -- because we really mean to capture the interval from 1 hour before Y right up until (and including) the _millisecond before_ Y.  Would you agree?

**Comment by brynrhodes** [comment]
So, I think that `X 1 hour or less before Y` _would_ be `true` if `X` is at the exact same time as `Y` (because I said an hour or less, and 0 hours is less than 1 hour). The more that I think about this case, the more that I think it overrides the exclusive definition of the before. In fact, I think it's even broader than I initially proposed above; consider `X 1 hour before Y`. Intuitively, if `X` is exactly 1 hour before `Y`, it should return true, but based on the exclusive definition of before, it should return false.

In other words, I think the relative qualifiers should override the exclusivity of the before/after comparison operator in all cases, not just where the two boundaries meet. Thoughts?

**Comment by cmoesel** [comment]
I'm afraid I don't agree that if `X` is the same as `Y`, that `X 1 hour or less before Y` _should_ be true.  I see `1 hour or less` as a qualifier of `X before Y` ("how much before Y?").  It shouldn't change the general relationship of `X before Y`.  If you consider it a qualifier, then it follows that the set of `X` that fulfills `X 1 hour or less before Y` should be a _subset_ of the set of `X` that fulfills `X before Y`.  If we allow `X same as Y`, then that's not true.

I probably come with this set of assumptions because this is how QDM execution of timing relationships works.  _First_ it checks that the general relationship is true (`X before Y`) and _then_ it checks that the delta fits within the specified range.  This is described in section 3.8 of the QDM 4.3 specification:

> Note that when filtering out data elements that do not meet temporal constraints (Step 6), first the timing relationship is evaluated using minute granularity and then the delta (if specified) is evaluated using the granularity of the specified timing unit. For example, `< 3 day(s) starts before start of` would first evaluate if the `starts before start of` constraint is met. If so, it would then determine if the delta is less than 3 days.

In QDM, if you wanted to include the RHS (instead of excluding it), there is a special `or concurrent with` construct for that: `< 3 days(s) start before or concurrent with start of`.

**Comment by brynrhodes** [comment]
So, I agree there ought to be a way to say `X before or at Y`, but I see that as a different issue, and I think it's both unintuitive and unnecessarily complicated to use that mechanism to get the desired behavior here; that would mean you'd have to say: 'X 1 hour or less before or at Y`, which feels pretty unnatural to me. If I say, I'll pick you up an hour or less before 8:00, would you be surprised if I showed up at 8:00 exactly?

**Comment by cmoesel** [comment]
I might not be surprised if you showed up at 8:00, but I'd consider you late since you said you'd be here before 8:00. ;-)

I think that the expectation of things really depends on the use case.  But I expect there are many cases where authors really want to ensure the _order_ of things, and establish that X did occur _before_ y within a certain window.  For example, I want to ensure that a patient was diagnosed with `X` 6 weeks or less before Encounter `Y` -- if the patient is diagnosed _at_ Encounter `Y`, then the CQM/CDS does not apply.  If we make the `X 6 weeks or less before Y` _include_ `Y`, then an author would have to say:  `X before Y and X 6 weeks or less before Y`.  That's awkward as well (seemingly redundant to me).

Clearly there is room for interpretation here -- and we need to do whatever is most consistent with the rest of CQL and least surprising to authors.  I just happen to think that overriding an exclusive boundary is less consistent w/ the spec, and that it will be surprising to QDM authors who are familiar w/ how QDM works.  But perhaps it would make sense to bring this up with real authors who will use the spec?

This is a total aside, but it also strikes me that the current way ELM does duration (by counting _whole_ units of time) almost implies that the correct phrase would be "6 weeks or fewer" (since it's countable units) rather than "6 weeks or less" (which implies more of a continuum).  But now I'm just getting nerdy.

**Comment by brynrhodes** [comment]
I didn't say "before" though, I said "an hour or less before" :)

I agree it would definitely be awkward to have to state the before again in the case where you really do want exclusive behavior.

I agree it needs discussion with the authors, it's on the agenda already for an upcoming meeting. I'll bring the results of that discussion back here.

**Comment by brynrhodes** [comment]
Quite apart from whether the expression returns true for events occurring at the same time, the result that an event occurring 30 minutes before doesn't return true either is pretty surprising. So it seems worth revisiting, should "X 1 hour or less before Y" return true if X is 30 minutes before Y? You can get that behavior by saying "X 1 hour or less before minute of Y", of course, but then, what about if X is 30 seconds before Y? Thoughts?

**Comment by cmoesel** [comment]
Oh -- I totally agree that "X 1 hour or less before Y" _should_ return `true` when X is 30 minutes before Y.  The way it currently works is definitely surprising.  I think it is because we think of time as a continuum -- so when I say "less than an hour", I mean _every single moment_ less than an hour.

**Comment by brynrhodes** [comment]
So, looking through the specified interpretations, I think this can be corrected quite simply by saying the timing phrases should be interpreted in terms of comparison operators on the boundaries, rather than by using a duration calculation. Do you see any issues with that approach?

**Comment by cmoesel** [comment]
So do you mean that `X 1 hour or less before Y` would be translated as `X same as or after (Y - 1 hour) and X before Y` -- or perhaps even `X in [Y - 1 hour, Y)`?

**Comment by brynrhodes** [comment]
Yes, that's what I was thinking, and the in formulation is just a shorthand for the equivalent comparisons.

**Comment by brynrhodes** [comment]
On this issue, feedback from the focus group today was that the before should always be interpreted exclusively and that there should be an additional "or on"/"on or" qualifier added to support the case where an inclusive boundary is needed. So the same as the "or concurrent with" qualifier in QDM (but hopefully a little more concise.

**Comment by cmoesel** [comment]
What does the inclusive variety look like in the context of a full expression?  For example, how do I make `X 1 hour or less before Y` inclusive?  Also, was using "or same as"/"same as or" considered for consistency with existing CQL timing operators?

**Comment by brynrhodes** [comment]
Sorry for the delay, so.... it looks like:

    X 1 hour or less on or before Y

Yes, we considered `same as` for consistency with the comparison phrasing, but it's very unnatural:

    X 1 hour or less same or before Y

No one would ever say that or intuit what it means.

**Comment by brynrhodes** [comment]
CDS WG and measure developers all agreed with the change and it was approved as STU comment 1337 on Wednesday's call. The change has been applied to the specification and the translator.
